crates/dbsp/src/operator/sample.rs

```diff
@@ -0,0 +1,222 @@
+//! Compute random samples of data.
+
+use crate::{
+    algebra::{HasOne, ZRingValue},
+    circuit::{
+        operator_traits::{BinaryOperator, Operator},
+        Scope,
+    },
+    trace::{Batch, BatchReader, Builder},
+    Circuit, DBData, DBWeight, OrdZSet, RootCircuit, Stream,
+};
+use rand::thread_rng;
+use std::{borrow::Cow, marker::PhantomData};
+
+// TODO: Operator to randomly sample `(K, V)` pairs.  This is
+// a little more tricky and also more expensive to implement than
+// sampling keys, as at the low level we need to pick random
+// values (`V`) from `OrderedLayer` and map the back to key (`K`)
+// by doing a lookup in the offset array.
+
+impl<B> Stream<RootCircuit, B>
+where
+    B: BatchReader<Time = ()> + Clone,
+    B::R: ZRingValue,
+{
+    /// Generates a uniform random sample of keys from `self`.
+    ///
+    /// At every clock tick, computes a random sample of the input batch
+    /// using [`BatchReader::sample_keys`].  The `sample_size` stream
+    /// specifies the size of the sample to compute (use `0` when no sample
+    /// is needed at the current clock cycle to make sure the operator
+    /// doesn't waste CPU cycles).
+    ///
+    /// Outputs a Z-set containing randomly sampled keys with weight `1`.
+    ///
+    /// This is not an incremental operator.  It samples the input
+    /// batch received at the current clock cycle and not the integral
+    /// of the input stream.  Prefix the call to `sample_keys()` with
+    /// `integrate_trace()` to sample the integral of the input.
+    ///
+    /// WARNING: This operator (by definition) returns non-dererministic
+    /// outputs.  As such it may not play well with most other DBSP operators
```

Comment:
Could determinism be ensured by using a RNG seeded by, for example, the clock cycle number?

Reply:
In hillview we made the server-side completely deterministic by having the client send the random seed.
I think the solution is to treat the seed similar to times: the seed values come in a (hidden) stream.

Reply:
Unfortunately, that won't make it deterministic.  The second source of non-determinism is the layout of batches in the trace, which depends on the amount of compaction work that has been done on it.

Reply:
> Unfortunately, that won't make it deterministic. The second source of non-determinism is the layout of batches in the trace, which depends on the amount of compaction work that has been done on it.

Oh, hmm, and I suppose the amount of compaction work is itself nondeterministic? Interesting.

I think it would be worth documenting that source of nondeterminism in the doc comment here. It didn't occur to me.